src/Processors/QueryPlan/PartsSplitter.cpp

```diff
@@ -961,33 +962,81 @@ RangesInDataParts findPKRangesForFinalAfterSkipIndexImpl(RangesInDataParts & ran
+
+    for (size_t part_index = 0; part_index < ranges_in_data_parts.size(); ++part_index)
+    {
+        const auto & index_granularity = ranges_in_data_parts[part_index].data_part->index_granularity;
+        const auto & part_lower_bound = index_access.getValue(part_index, 0);
+        const auto & part_upper_bound = index_access.getValue(part_index, index_granularity->getMarksCountWithoutFinal());
+        if ((compareValues(selected_lower_bound.value, part_upper_bound, false) > 0) ||
+            (compareValues(selected_upper_bound.value, part_lower_bound, false) < 0))
+        {
+            continue; /// early exit, intersection infeasible in this part
         }
 
-        for (size_t range_begin = 0; range_begin < is_selected_range.size(); range_begin++)
+        std::vector<PartsRangesIterator> part_candidate_ranges;
+        for (size_t range_begin = 0; range_begin < index_granularity->getMarksCountWithoutFinal(); range_begin++)
         {
             const bool value_is_defined_at_end_mark = ((range_begin + 1) < index_granularity->getMarksCount());
             if (!value_is_defined_at_end_mark)
             {
                 return skip_and_return_all_part_ranges();
             }
 
-            if (is_selected_range[range_begin])
+            if (part_selected_ranges[part_index].find(range_begin) != part_selected_ranges[part_index].end())
                 continue;
             MarkRange rejected_range(range_begin, range_begin + 1);
-            rejected_ranges.push_back(
-                {index_access.getValue(part_index, rejected_range.begin), false, rejected_range, part_index, PartsRangesIterator::EventType::RangeStart, false});
+            part_candidate_ranges.push_back(
+                {index_access.getValue(part_index, rejected_range.begin), false, rejected_range, part_index,
+                    PartsRangesIterator::EventType::RangeStart, false});
         }
+        /// add the key of the final mark for setting the right upper bound of the part
+        MarkRange final_range(index_granularity->getMarksCountWithoutFinal(), index_granularity->getMarksCountWithoutFinal() + 1);
+        part_candidate_ranges.push_back(
+            {index_access.getValue(part_index, final_range.begin), false, final_range, part_index,
+                PartsRangesIterator::EventType::RangeStart, false});
+
+        /// Optimization - We don't add all ranges of the part, but only the ranges inside the feasible
+        /// intersection boundary.
+        auto candidates_start = std::upper_bound(part_candidate_ranges.begin(), part_candidate_ranges.end(),
+                                                 selected_lower_bound);
+        if (unlikely(candidates_start == part_candidate_ranges.end()))
+            continue; /// no intersection possible in this part
+        auto candidates_end = std::upper_bound(part_candidate_ranges.begin(), part_candidate_ranges.end(),
```

Comment:
`IndexAccess` has methods like `findRightmostMarkLessThanValueInRange` that seem to do something very similar - can we use them?